File: src/main/java/org/sonar/plugins/xquery/checks/AbstractProhibitVariableCheck.java
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks;

import org.apache.commons.lang.StringUtils;
import org.sonar.plugins.xquery.parser.XQueryParser;
import org.sonar.plugins.xquery.parser.XQueryTree;

import java.util.regex.Pattern;

public abstract class AbstractProhibitVariableCheck extends AbstractCheck {
      
    @Override
    public void enterExpression(XQueryTree node) {
        super.enterExpression(node);
        
        switch (node.getType())
        {
           case XQueryParser.ParamName:
           case XQueryParser.VarDecl:
           case XQueryParser.ForName:
           case XQueryParser.ForAt:
           case XQueryParser.LetName:
        	   String name = node.getValue("QName");
        	   if (!StringUtils.isBlank(name) && Pattern.matches(getPattern(), name)) {
        		   createIssue(node.getLine());
        	   }
        	   break;
        }
    }

    abstract protected void createIssue(int lineNumber);
    abstract protected String getPattern();
}

File: src/main/java/org/sonar/plugins/xquery/checks/AbstractProhibitVariableCheck.kt
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks

import org.antlr.v4.runtime.ParserRuleContext
import org.apache.commons.lang.StringUtils
import org.sonar.plugins.xquery.parser.*
import org.sonar.plugins.xquery.parser.XQueryParser.*

import java.util.ArrayList
import java.util.regex.Pattern

abstract class AbstractProhibitVariableCheck : AbstractCheck() {

    protected abstract val pattern: String

    override fun enterExpression(node: ParserRuleContext) {
        super.enterExpression(node)
        val names = ArrayList<String>()

        when (node) {
            is ForVarContext -> {
                names.add(node.name.text)
                names.add(node.pvar.text)
//                names.add(node.getValue("qName"))
            }
            is ParamContext -> names.add(node.name.text)
            is VarDeclContext -> names.add(node.name.text)
            is LetVarContext -> names.add(node.name.text)
        }

        if (names.stream()
                .filter { it.isNotBlank() }
                .filter { name -> Pattern.matches(pattern, name) }
                .count() > 0) {
            createIssue(node.getLine())
        }

    }

    protected abstract fun createIssue(lineNumber: Int)
}
